File: src/Muflone/CustomTypes/When.cs
using Newtonsoft.Json;
using System;

namespace Muflone.CustomTypes;

public record When
{
    private readonly long _ticks;

    public DateTime Value => new(_ticks);

    public When(DateTime dateTime)
    {
        _ticks = dateTime.Ticks;
    }

    [JsonConstructor]
    public When(long ticks)
    {
        _ticks = ticks;
    }
}

File: php/src/Muflone/CustomTypes/When.php
<?php

declare(strict_types=1);

namespace Muflone\CustomTypes;

use DateTimeImmutable;

final readonly class When
{
    private int $microseconds;

    public function __construct(DateTimeImmutable|int $value)
    {
        if ($value instanceof DateTimeImmutable) {
            $this->microseconds = (int)($value->getTimestamp() * 1000000 + (int)$value->format('u'));
        } else {
            $this->microseconds = $value;
        }
    }

    public function getValue(): DateTimeImmutable
    {
        $timestamp = intdiv($this->microseconds, 1000000);
        $microseconds = $this->microseconds % 1000000;

        $dateTime = DateTimeImmutable::createFromFormat('U', (string)$timestamp);
        if ($dateTime === false) {
            $dateTime = new DateTimeImmutable('@' . $timestamp);
        }

        return $dateTime->modify(sprintf('+%d microseconds', $microseconds));
    }

    public function getMicroseconds(): int
    {
        return $this->microseconds;
    }
}
